docs/tutorial/source-plugin-tutorial.md

```diff
@@ -0,0 +1,984 @@
+
+You can read about [how to use Gatsby Image to prevent image bloat](/docs/using-gatsby-image/) if you are unfamiliar with it.
+
+#### Create `remoteFilenNode`'s from a URL
+
+To create optimized images from URLs, `File` nodes for image files need to be added to your site's data. Then, `gatsby-plugin-sharp` and `gatsby-transformer-sharp` need to be installed, and they will automatically find image files and add the data needed for `gatsby-image`.
+
+Start by installing `gatsby-source-filesystem` in the `source-plugin`:
+
+```shell:title=source-plugin
+npm install gatsby-source-filesystem
+```
+
+Now in your plugin's `gatsby-node`, you can implement a new API that gets called everytime a node is created called `onCreateNode`. You can check if the node created was one of your `Post` nodes, and if it was, create a file from the URL on the `imgUrl` field.
+
+Import the `createRemoteFileNode` helper from `gatsby-source-filesystem`, which will download a file from a remote location and create a `File` node for you.
+
+```javascript:title=source-plugin/gatsby-node.js
+const { ApolloClient } = require("apollo-client")
+const { InMemoryCache } = require("apollo-cache-inmemory")
+const { split } = require("apollo-link")
+const { HttpLink } = require("apollo-link-http")
+const { WebSocketLink } = require("apollo-link-ws")
+const { getMainDefinition } = require("apollo-utilities")
+const fetch = require("node-fetch")
+const gql = require("graphql-tag")
+const WebSocket = require("ws")
+// highlight-start
+const { createRemoteFileNode } = require(`gatsby-source-filesystem`)
+// highlight-end
+```
+
+Then export a new function `onCreateNode`, and call `createRemoteFileNode` in it whenever a node of of type `Post` is created:
+
+```javascript:title=source-plugin/gatsby-node.js
+// called each time a node is created
+exports.onCreateNode = async ({
+  node, // the node that was just created
+  actions: { createNode },
+  createNodeId,
+  getCache,
+}) => {
+  if (node.internal.type === POST_NODE_TYPE) {
+    const fileNode = await createRemoteFileNode({
+      // the url of the remote image to generate a node for
+      url: node.imgUrl,
+      parentNodeId: node.id,
+      createNode,
+      createNodeId,
+      getCache,
+    })
+
+    if (fileNode) {
+      node.remoteImage___NODE = fileNode.id
+    }
+  }
+}
+```
+
+This code is called every time a node is created, i.e. when `createNode` is invoked. Each time it is called in the `sourceNodes` step, the condition will check if the node was a `Post` node since those are the only nodes with an image associated with them in your case. Then a remote node is created, if it's successful, the `fileNode` is returned. The next few lines are important:
```

Comment:
```suggestion
This code is called every time a node is created, e.g. when `createNode` is invoked. Each time it is called in the `sourceNodes` step, the condition will check if the node was a `Post` node. Since those are the only nodes with an image associated with them, that is the only time images need to be optimized. Then a remote node is created, if it's successful, the `fileNode` is returned. The next few lines are important:
```